backend.native/tests/runtime/workers/freeze6.kt

```diff
@@ -30,5 +34,126 @@ fun ensureNeverFrozenFailsTarget(){
     println("OK")
 }
 
-data class Hi(val s:String)
-data class Nested(val hi:Hi)
\ No newline at end of file
+fun createRef1(): FreezableAtomicReference<Any?> {
+    val ref = FreezableAtomicReference<Any?>(null)
+    ref.value = ref
+    ref.freeze()
+    ref.value = null
+    return ref
+}
+
+var global = 0
+
+@Test
+fun ensureFreezableHandlesCycles1() {
+    val ref = createRef1()
+    kotlin.native.internal.GC.collect()
+
+    val obj: Any = ref
+    global = obj.hashCode()
+}
+
+class Node(var ref: Any?)
+
+/**
+ *    ref1 -> Node <- ref3
+ *             |        /\
+ *             V        |
+ *             ref2  ---
+ */
+fun createRef2(): FreezableAtomicReference<Node?> {
+    val ref1 = FreezableAtomicReference<Node?>(null)
+
+    val node = Node(null)
+    val ref3 = FreezableAtomicReference<Any?>(node)
+    val ref2 = FreezableAtomicReference<Any?>(ref3)
+
+    node.ref = ref2
+    ref1.value = node
+
+    ref1.freeze()
+    ref3.value = null
+
+    assertTrue(node.isFrozen)
+    assertTrue(ref1.isFrozen)
+    assertTrue(ref2.isFrozen)
+    assertTrue(ref3.isFrozen)
+
+    return ref1
+}
+
+@Test
+fun ensureFreezableHandlesCycles2() {
+    val ref = createRef2()
+    kotlin.native.internal.GC.collect()
+
+    assertTrue(ref.value.toString().length > 0)
```

Comment:
Not sure if checking that `ref1` is not corrupted makes sense here, because it wouldn't be a part of condensation in any case.